README.md
# IGD: A high-performance search engine for large-scale genomic interval datasets

## Summary

One type of genome analysis is to compare a set of genomic loci (a region set) to many other annotated region sets. A wide variety of computational tools, such as LOLA, BART, Bedtools, Tabix, GenomeRunner and Giggle, have been developed for this purpose. Among these tools, Giggle claims to be over three orders of magnitude faster than others in searching millions to billions of genomic intervals. **IGD**, an integrated genome database, takes a database approach that directly integrates large-scale data sources into an easily searchable database by reshaping the data records. As a database, IGD takes not only the genomic regions, but also the signal levels and/or strand information, which makes parameterized query (dynamic search) an easy task. The search speed of IGD is one to two orders of magnitude faster than Giggle while the database size is several times smaller than that of the Giggle index files. 

## Introduction

Genome data sources, such as Roadmap, UCSC, and Cistrome, contain thousand of annotated data sets and each dataset has thousands to millions of genomic regions. 
The enrichment analysis is to find overlaps between regions of all data sets in the data source and all regions in the query data set. This becomes challenging when the size of the data source regions and the size of the query regions are large. For example, UCSC human data source contains about 7 billion regions or intervals, so if the query set has 100,000 intervals, then a total of 700 trillion comparisons may be involved. To effectively search for the overlaps, different indexing techniques have been developed. For example, Giggle employs a B+ tree to build many pure indexing files, and then a query is carried out on these indexing files instead of on the original data. Using the indexing, the actual search space is significantly reduced. 

The goal of IGD is to build a database that integrates all of the genomic data sets from one or more data sources and minimizes the actual search space for a general query. To achieve this, IGD reshapes the data sources by dividing the genome into a large number of equal-size bins (around 200,000 bins) and arranging data records of the data sources into the bin or bins they intersect. Data in each bin will be saved as a block of a single file. The file head contains the location of the bins and the number of IGD data elements in each bin. Each IGD data element contains the index of the original dataset, the original genomic region (start and end coordinates), and a value (signal level and/or strand information). To find overlaps of a query, one only needs to load one or a few bin block data instead of all data sets, which minimizes the data loading time and the comparisons are carried out only in the loaded bins, which minimizes the actual search space. A link with details about the implementation will be provided later.

## How to build IGD

If zlib is not already installed, install it:
```
sudo apt-get install libpng12-0
```

Then:
```
git clone https://github.com/databio/IGD.git
cd IGD
make
```

the executable `igd` is in the subfolder `bin`. And then copy it to /usr/local/bin.

## How to run IGD

<<<<<<< HEAD
### 1. Create iGD database
 
#### 1.1 Create iGD database from a genome data source folder
=======
### 1. Create IGD database
 
#### 1.1 Create IGD database from a genome data source folder
>>>>>>> 42a3ce8e28829decf75d9ff19eef2dbf843ddee3
```
igd create "/path/to/data_source_folder/*" "/path/to/igd_folder/" "databaseName" [option]
```

where:
- "path/to/data_source_folder/" is the path of the folder that contains .bed.gz or .bed data files.
- "path/to/igd_folder/" is the path to the output igd folder;
- "databaseName" is the name you give to the database, for eaxmple, "roadmap"

option:

-b: bin-size (power of 2; default 14, which is 16384 bp)
```
#### 1.2 Create IGD database from a list of source files
 
```
igd create "/path/to/source-list file" "/path/to/igd_folder/" "databaseName" -f [option]
```

where:

- "/path/to/source-list file" is the path to the file that lists the source files
- "path/to/igd_folder/" is the path to the output igd folder;
- "databaseName" is the name you give to the database, for eaxmple, "roadmap"

option:

-b: bin-size (power of 2; default 14, which is 16384 bp)
```
#### 1.2 Create iGD database from a list of source files
 
```
igd create "/path/to/source-list file" "/path/to/igd_folder/" "databaseName" -f [option]

where:

- "/path/to/source-list file" is the path to the file that lists the source files

- "path/to/igd_folder/" is the path to the output igd folder;

- "databaseName" is the name you give to the database, for eaxmple, "roadmap"

option:

-b: bin-size (power of 2; default 14, which is 16384 bp)
```



### 2. Search IGD for overlaps
```
igd search "path/to/query_file" "path/to/igd_data_file" [options]
```

where:
- path/to/query_file is the path to the query file
- path/to/igd_data_file is the path to the igd data (mode 1)

options:
-v: dynamic search with threshold signal value (0-1000)
-o: output file-name

For a detailed example, please check out the `vignettes`.

<<<<<<< HEAD
## iGD databases
Downloads of some iGD databases (fully created and directly searchable by using iGD) are available at [big.databio.org/example_data/igd](http://big.databio.org/example_data/igd).

## R-wrapper of IGD

### 1. Create iGD database 
#### 1.1  from a genome data source
```
> library(IGDr)
> createIGD("/path/to/data_source_folder/*" "/path/to/igd_folder/" "databaseName" [option]

where:

- "path/to/data_source_folder/" is the path of the folder that contains .bed.gz or .bed data files.

- "path/to/igd_folder/" is the path to the output igd folder;

- "databaseName" is the name you give to the database, for eaxmple, "roadmap"

options:

-b: bin size in bp (default 16384)
```
#### 1.2  from a file that contains the list of genome data source files 
```
> library(IGDr)
> createIGD_f("/path/to/source-list file" "/path/to/igd_folder/" "databaseName" [option]

where:

- "path/to/the list file/" is the path to the file that contains the .bed.gz or .bed data files.

- "path/to/igd_folder/" is the path to the output igd folder;

- "databaseName" is the name you give to the database, for eaxmple, "roadmap"

options:

-b: bin size in bp (default 16384)
```

### 2. search the igd database in R (an example for a created igd file)

Search the igd database with a single query:
```
> igd_file = "igdr_b14/roadmap.igd"
> library(IGDr)
> igd <- IGDr::IGDr(igd_file)
> hits <- search_1r(igd, "chr6", 1000000, 10000000)
> hits
```
Search the igd database with n queries:
```
> igd_file = "igdr_b14/roadmap.igd"
> library(IGDr)
> igd <- IGDr::IGDr(igd_file)
> chrms = c("chr6", "chr1", "chr2")
> starts = c(10000, 100000, 1000000)
> ends = (100000, 1000000, 10000000)
> hits <- search_nr(igd, 3, chrms, starts, ends)
> hits
```
Search a whole query file chainRn4.bed
```
> igd_file = "igdr_b14/roadmap.igd"
> query_file = "r10000.bed"
> library(bit64)
> library(IGDr)
> fi = IGDr::getFInfo(igd_file)
> hits = integer64(fi$nFiles)
> ret = IGDr::search_all(igd_file, query_file, hits)
> for(i in 1:fi$nFiles){
  cat(i, "\t", toString(ret[i]), "\t", toString(fi$fInfo[i,2]), "\n")
  }
>
```
=======
## IGD database downloads

Pre-built IGD databases can be downloaded from [big.databio.org/example_data/igd](http://big.databio.org/example_data/igd).

## R interface to IGD

You can find an R package that uses the IGD C library at [http://github.com/databio/rigd](github.com/databio/rigd).
>>>>>>> 42a3ce8e28829decf75d9ff19eef2dbf843ddee3


changelog.md
# Changelog

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format. 

## [0.0.1] -- Unreleased
### Added
- lorem ipsum

### Changed
- lorem ipsum

### Fixed
- lorem ipsum

## [0.0.0] -- 2020-04-08

First release


src/igd_t2w.c
//=====================================================================================
//Read igd region data and query data, and then find all overlaps 
//by Jianglin Feng  05/12/2018
//igd_tiles2w: put all tiles data into a single file 
//-------------------------------------------------------------------------------------
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <err.h>
#include <time.h>
#include <math.h>
#include <float.h>
#include <glob.h>
#include <errno.h>

//-------------------------------------------------------------------------------------
struct igd_data
{
    uint32_t i_idx;        			//genomic object--data set index
    uint32_t r_start;      			//region start
    uint32_t r_end;        			//region end
    uint32_t g_val;        			//signal level
};

char *fileBase = "_b14_";         	//14 bits block
uint32_t nmax[] = {15940, 15580, 12760, 12240, 11670, 10990, 10260, 9370, 8860, 8610, 8710, 
        8580, 7300, 6840, 6510, 5830, 5370, 5160, 3820, 4150, 2980, 3240, 9880, 3510};
char *folder[] = {"chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", 
        "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", 
     	 "chr17", "chr18", "chr19", "chr20", "chr21", "chr22", "chrX", "chrY"};
uint32_t gstart[25] = {0, 15940, 31520, 44280, 56520, 68190, 79180, 89440, 98810, 107670, 
        116280, 124990, 133570, 140870, 147710, 154220, 160050, 165420, 170580, 174400, 
        178550, 181530, 184770, 194650, 198160}; 
uint32_t nTiles = 198160;
uint32_t nbp = 16384, bgz_buf = 1024;
uint64_t maxCount = 134217728;//maxCount*16=2GB,67108864,33554432, 67108864;//
uint32_t *g2ichr;

//-------------------------------------------------------------------------------------
int compare_iidx(const void *a, const void *b)
{
    struct igd_data *pa = (struct igd_data *) a;
    struct igd_data *pb = (struct igd_data *) b;
    return pa->i_idx - pb->i_idx;
}

int compare_rend(const void *a, const void *b)
{
    struct igd_data *pa = (struct igd_data *) a;
    struct igd_data *pb = (struct igd_data *) b;
    return pa->r_end - pb->r_end;
}

//reload tile igd files, sort them and save them into a single file
void tiles2w(char *tPath, char* oPath, char *igdName)
{   //tPath: for igd tile data files
    //oPath: for the single igd file 
    int i, k;
    char idFile[256];
    sprintf(idFile, "%s%s%s", oPath, igdName, ".igd");    
    FILE *fp1 = fopen(idFile, "wb"); 
    if(fp1==NULL)
        printf("Can't open file %s", idFile);     
    uint32_t nrec;  
    FILE *fp0; 
    
    clock_t start, end;
    start = clock();        
    //struct igd_data *gdata;
    uint32_t *counts = calloc(nTiles, sizeof(uint32_t)); 
    fwrite(counts, sizeof(uint32_t), nTiles, fp1);

    char iname[256]; 
    struct igd_data *gdata; 
    printf("Starting ...\n");         
    for(i=0;i<nTiles;i++){
        k = g2ichr[i];
        sprintf(iname, "%s%s%s/%s_%u%s", tPath, "data0/", folder[k], igdName, i-gstart[k], ".igd");
        fp0 = fopen(iname, "rb");
        if(fp0!=NULL){   
            fseek(fp0, 0, SEEK_END);
            nrec = ftell(fp0)/sizeof(struct igd_data);
            fseek(fp0, 0, SEEK_SET);
            if(nrec>0){
                gdata = malloc(nrec*sizeof(struct igd_data));
                fread(gdata, sizeof(struct igd_data), nrec, fp0);
                fclose(fp0);
                qsort(gdata, nrec, sizeof(struct igd_data), compare_rend);
                //append the data to fp1
                fwrite(gdata, sizeof(struct igd_data), nrec, fp1);
                counts[i] = nrec;
                free(gdata);
                //gdata = NULL;
            }      
            //fclose(fp0);
        }
        if(i%1000==0)
            printf("%u\t%u\t%f \n", i, nrec, (double)(clock()-start)/CLOCKS_PER_SEC);
    }
    fseek(fp1, 0, SEEK_SET);
    fwrite(counts, sizeof(uint32_t), nTiles, fp1);
    fclose(fp1); 
    end = clock();    
    free(counts);   
    printf("igd_w finished: time: %f \n", (double)(end-start)/CLOCKS_PER_SEC);    
}

//-------------------------------------------------------------------------------------
int main(int argc, char **argv)
{
    if (argc < 4) 
        errx(1, "usage:\t%s <tile-file path><ouput path><dbName>\n", argv[0]);   
    //convert block index to chr index for convenience
    g2ichr = malloc(nTiles*sizeof(uint32_t));
    uint32_t i, j;
    for(i=0; i<24; i++){  
        	for(j=gstart[i]; j<gstart[i+1]; j++)      
        	    g2ichr[j] = i;
    }
    char *tpath = argv[1];
    char *opath = argv[2];
    char *dbname = argv[3];
    tiles2w(tpath, opath, dbname);    

    free(g2ichr);
}

/*
echo cfq > /sys/block/hda/queue/scheduler

*/